receiver/receivercreator/hints.go

```diff
@@ -0,0 +1,178 @@
+type hintsTemplatesBuilder interface {
+	createReceiverTemplateFromHints(env observer.EndpointEnv) (*receiverTemplate, error)
+}
+
+// k8sHintsBuilder creates configurations from hints provided as Pod's annotations.
+type k8sHintsBuilder struct {
+	logger          *zap.Logger
+	config          DiscoveryConfig
+	ignoreReceivers map[string]bool
+}
+
+var _ hintsTemplatesBuilder = (*k8sHintsBuilder)(nil)
+
+func createK8sHintsBuilder(config DiscoveryConfig, logger *zap.Logger) k8sHintsBuilder {
+	ignoreReceivers := make(map[string]bool, len(config.IgnoreReceivers))
+	for _, r := range config.IgnoreReceivers {
+		ignoreReceivers[r] = true
+	}
+	return k8sHintsBuilder{
+		logger:          logger,
+		config:          config,
+		ignoreReceivers: ignoreReceivers,
+	}
+}
+
+// createReceiverTemplateFromHints creates a receiver configuration based on the provided hints.
+// Hints are extracted from Pod's annotations.
+// Scraper configurations are only created for Port Endpoints.
+// TODO: Log receiver configurations are only created for Pod Container Endpoints.
+func (builder *k8sHintsBuilder) createReceiverTemplateFromHints(env observer.EndpointEnv) (*receiverTemplate, error) {
+	var endpointType string
+	var pod observer.Pod
+
+	endpointType = getStringEnv(env, "type")
+	if endpointType == "" {
+		return nil, fmt.Errorf("could not get endpoint type: %v", zap.Any("env", env))
+	}
+
+	if endpointType != string(observer.PortType) {
+		return nil, nil
+	}
+
+	builder.logger.Debug("handling hints for added endpoint", zap.Any("env", env))
+
+	if endpointPod, ok := env["pod"]; ok {
+		err := mapstructure.Decode(endpointPod, &pod)
+		if err != nil {
+			return nil, fmt.Errorf("could not extract endpoint's pod: %v", zap.Any("endpointPod", pod))
+		}
+		if pod.Annotations != nil {
+			if len(pod.Annotations) == 0 {
+				return nil, nil
+			}
+		}
+	} else {
+		return nil, nil
+	}
+
+	return builder.createScraper(pod.Annotations, env)
+}
+
+func (builder *k8sHintsBuilder) createScraper(
+	annotations map[string]string,
+	env observer.EndpointEnv) (*receiverTemplate, error) {
+
+	var port uint16
+	var p observer.Port
+	err := mapstructure.Decode(env, &p)
+	if err != nil {
+		return nil, fmt.Errorf("could not extract port event: %v", zap.Any("env", env))
+	}
+	if p.Port == 0 {
+		return nil, fmt.Errorf("could not extract port: %v", zap.Any("env", env))
+	}
+	port = p.Port
+	pod := p.Pod
+
+	if !discoveryMetricsEnabled(annotations, otelMetricsHints, fmt.Sprint(port)) {
+		return nil, nil
+	}
+
+	subreceiverKey := getHintAnnotation(annotations, otelMetricsHints, scraperHint, fmt.Sprint(port))
+	if subreceiverKey == "" {
+		// no scraper hint detected
+		return nil, nil
+	}
+	if _, ok := builder.ignoreReceivers[subreceiverKey]; ok {
+		// scraper is ignored
+		return nil, nil
+	}
+	builder.logger.Debug("handling added hinted receiver", zap.Any("subreceiverKey", subreceiverKey))
+
+	defaultEndpoint := getStringEnv(env, "endpoint")
```

Comment:
```suggestion
	defaultEndpoint := getStringEnv(env, endpointConfigKey)
```

Wdyt of using the global variable instead?

Reply:
Good point, actually we can re-use the one from https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/3ca96862b9205e447cf56c855505c714202cca17/receiver/receivercreator/config.go#L20.